### PR TITLE
Lra 558 lsa ride share add terms interface

### PR DESCRIPTION
**Pull request by britaumich**
The Terms link is added to the header.
There are Index, Show, New, and Edit Terms pages in the pull request.
Terms scopes are edited.
Current term(s): The term is current until the next term starts. In May-August there are two current terms.

### CODE REVIEW

File: app/models/term.rb
Comment (dschmura) [inline app/models/term.rb:1]: This bit is a little difficult to read 

` scope :current, -> { all.where(':date BETWEEN term_start AND term_end', date: Date.today).or(all.where(term_start: all.where('term_start < :date', date: Date.today).max.term_start)) }`

I've been trying to write it out in pseudo code and I keep getting lost. 

The current term is 
- all terms for which the date submitted falls between the term_start and term_end dates
- OR
- all of the terms that have a term_start date that is before the date submitted.

Am I understanding this correctly?

Comment (britaumich) [reply app/models/term.rb:1]: Almost correctly. Look at the **MAX** word below.
The current term is

- all terms for which the date submitted falls between the term_start and term_end dates
  - **Yes**

OR
- all of the terms that have a term_start date that is before the date submitted.
  - all the terms that have a **MAX** (of all the term_start date that is before the date submitted)

If you can write a better query - that would be great!

Comment (britaumich) [reply app/models/term.rb:1]: @dschmura Please look at my last commit [bdcc023](https://github.com/lsa-mis/lsa_ride_share/pull/43/commits/bdcc0231abb8059482cdbbe9dcfc497606d5252a)
What do you think about these scopes?

All the following tests work:
Here are tests for different dates:
d = Date.today (Fri, 21 Apr 2023) - current.name = "Winter 2023"
d = Date.today + 1.month (Sun, 21 May 2023) current.name = ["Spring/Summer 2023", "Spring 2023"]
d = Date.today + 3.month (Fri, 21 Jul 2023) current.name = ["Spring/Summer 2023", "Summer 2023"]
d = Date.today + 5.month (Thu, 21 Sep 2023) current.name = "Fall 2023"

Comment (dschmura) [inline app/models/term.rb:1]: Nice! I think this is much clearer.

**Review by dschmura — APPROVED** [review]
Nice refactor!

**Comment by britaumich** [comment]
I added three more commits to the PR. Please review.

**Review by dschmura — APPROVED** [review]

File: app/models/term.rb [hunk 1]
```diff
@@ -11,7 +11,13 @@
 #  updated_at :datetime         not null
 #
 class Term < ApplicationRecord
-  
-  scope :current, -> { where(':date BETWEEN term_start AND term_end', date: Date.today)}
+
+  validates_presence_of :code, :name, :term_start, :term_end
+
```
Comment (anantaUM) [inline app/models/term.rb:16]: Please consider renaming _term_start_ and _term_end_ to _term_start_date_ and _term_end_date_ respectively. Makes it more clear to anyone reading this code.

File: app/models/term.rb [hunk 2]
```diff
@@ -11,7 +11,13 @@
...
+  scope :sorted, -> { where.not(term_start: nil).order(:term_start, term_end: :desc) }
+  scope :between_start_and_end_dates, -> { sorted.where(':date BETWEEN term_start AND term_end', date: Date.today) }
+  scope :started_before_today, -> { sorted.where('term_start < :date', date: Date.today) }
+  scope :max_started_before_today, -> { sorted.where(term_start: started_before_today.max.term_start) }
```
Comment (anantaUM) [inline app/models/term.rb:20]: I would like to make a suggestion here. Please correct me if I am wrong. If the today's date is outside a term, then the current logic is to display programs of the recently completed term. I think in such a scenario, we should not display any programs and render text like "No programs for current date / term". The user can then select required term from the dropdown to view relevant programs.

Comment (britaumich) [reply app/models/term.rb:20]: We should ask Sarah and Sheri about that. They are going to use the application.

File: app/models/term.rb (hunk 1 above)
Comment (britaumich) [reply app/models/term.rb:16]: Maybe it makes more sense to call these fields "classes_begin" and "classes_end" like they are called in the Academic Calendar. I'll talk to Sarah and Sheri.

Comment (anantaUM) [reply app/models/term.rb:16]: Yes, even these names are good. I would still suggest that you consider "date" word in the end. Thanks for following up with customers.

File: app/models/term.rb (hunk 2 above)
Comment (anantaUM) [reply app/models/term.rb:20]: ok thanks.

File: app/models/term.rb (hunk 1 above)
Comment (britaumich) [reply app/models/term.rb:16]: I renamed fields in the terms table: term_start to classes_begin_date; term_end to classes_begin_date.

File: app/models/term.rb (hunk 2 above)
Comment (britaumich) [reply app/models/term.rb]: @anantaUM I returned the Term current scope to where(':date BETWEEN classes_begin_date AND classes_end_date', date: Date.today). 
I edited the Programs index page to display a message if the current date is between terms or if there are no programs for the selected term.
I talked to Sarah and Sheri about start and end terms dates. Sarah doesn't care and Sheri wants it to be classes begin and end.
They are both concerned about being able to access programs between terms. That is not a problem, they can always access all programs when they select a term.
Please review the PR again.

File: app/controllers/programs_controller.rb [hunk 3]
```diff
@@ -121,13 +121,13 @@ def set_program
...
     def set_terms_and_units
-      @terms = Term.all.order(:term_start)
+      @terms = Term.sorted
       @units = Unit.where(id: current_user.unit).order(:name)
     end
 
     # Only allow a list of trusted parameters through.
     def program_params
-      params.require(:program).permit(:active, :title, :term_start, :term_end, :term_id, :subject, :catalog_number, :class_section, 
+      params.require(:program).permit(:active, :title, :term_id, :subject, :catalog_number, :class_section,
```
Comment (anantaUM) [inline app/controllers/programs_controller.rb:130]: @britaumich - The new variable names are not included here. Please add them.

**Review by anantaUM — COMMENTED** [review]
Added comment in **programs_controller.rb**

File: app/controllers/programs_controller.rb (hunk 3 above)
Comment (britaumich) [reply app/controllers/programs_controller.rb:130]: @anantaUM The programs table doesn't have term's starts and end dates, it has term_id. It did have them when the table was initially created, and we did not have the terms table at that moment. But then I created a terms table and added term_id to the programs table.

**Comment by britaumich** [comment]
@anantaUM Please review this PR as soon as possible on Monday. Since I changed column names in the terms table, all my other branches are broken right now on my localhost because they use old names.

**Review by anantaUM — APPROVED** [review]

**Pull request merged**